src/server_concurrent.py

```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
+"""File contains a simple HTTP server."""
+from __future__ import unicode_literals
+# import socket
+import os
+import io
+from mimetypes import guess_type
+try:
+    from http.client import HTTPException
+except ImportError:
+    from httplib import HTTPException
+
+CRLF = '\r\n'
+ROOT = './webroot'
+
+
+def http_server(socket, address):
+    '''handles to the imcoming http connections on a separate greenlet'''
+    buffer_length = 16
+    message = b''
+    message_complete = False
+    while not message_complete:
+        part = socket.recv(buffer_length)
+        message += part
+        if len(part) < buffer_length:
+            message_complete = True
+
+    print(message)
+
+    try:
+        uri = parse_request(message)
+    except HTTPException as exception:
+        msg = response_error(exception)
+    try:
+        file_data_tuple = resolve_uri(uri)
+    except HTTPException as exception:
+        msg = response_error(exception)
+    try:
+        msg = response_ok(file_data_tuple)
+    # Change to a bear except to handle unknown this
+    except HTTPException:
+        msg = (response_error("500 Internal Server Error"))
+
+    socket.sendall(msg.encode("utf-8"))
```

Comment:
this line worries me a bit.  this means that you are expecting `msg` to be unicode here.  If the response includes something like an image file, then you should certainly not attempt to decode that to unicode at any point.  if the `response_error` and `response_ok` functions always return encoded bytestrings, then you don't have to worry about that.